### PR TITLE
mpb: init at 1.11.1

### PR DESCRIPTION
**Pull request by sheevy**
Added support for MPB ([source](https://github.com/NanoComp/mpb), [docs](https://github.com/NanoComp/mpb)).
MPB is standalone C++ library, with bindings to Scheme and Python. It's packaged in a bit of convoluted way, Scheme binding live in the same repository as MPB itself while but Python binding are part of PyMeep package. Please see first paragraph from the [docs](https://mpb.readthedocs.io/en/latest/Python_User_Interface/). 
The only goal for this PR is to add support for MBP within PyMeep, not to have support for for MPB in full generality. This means:
* Scheme binding is out of scope
* MPI in MPB is out of scope, because PyMeep only supports serial MBP [source](https://meep.readthedocs.io/en/latest/Build_From_Source/#mpb)
Additionally:
* `--with-inv-symmetry` is not enabled, because we want as general version of MPB as possible.
* It's not clear to me if `--with-hermitian-eps` should be specified, or is it on by default. Package in [AUR](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=mpb-git) does not use it, however the from [conda-forge](https://github.com/conda-forge/mpb-feedstock/blob/main/recipe/build.sh) does.

On testing, I don't know how to test it from MPB itself, however I added a simple `import meep.mpb` to checks in `python3Packages.meep`

I tested if things work by creating a temporary `shell.nix` with `mpb`, `python3Packages.meep` and `python` available. Not sure what's the proper way of testing, so would appreciate any help on this.

@sheepforce for visibility

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/mp/mpb/package.nix
Comment (sheepforce) [inline pkgs/by-name/mp/mpb/package.nix:52]:
```suggestion
  doCheck = true;
```
With enabled libctl, guile and perl all tests pass. If possible, having tests is a good idea.

File: pkgs/by-name/mp/mpb/package.nix
Comment (sheepforce) [inline pkgs/by-name/mp/mpb/package.nix:60]: This only has meaning if you add yourself to https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix
That's okay if you wish to maintain that package long term. In that case make a separate commit, where you add yourself to that file. If not, it's also fine if you keep just me here. I will have a look on that package.

File: pkgs/development/python-modules/meep/default.nix
Comment (sheepforce) [inline pkgs/development/python-modules/meep/default.nix:137]: If you are not using that module in the test script, it may be more clear to add `pythonImportsCheck = [ `meep.mpb` ];` instead of importing that one here.

File: pkgs/by-name/mp/mpb/package.nix
Comment (sheepforce) [inline pkgs/by-name/mp/mpb/package.nix:46]:
```suggestion
    "--with-libctl=yes"
    "--with-libctl=${libctl}"
```
You've libctl in the inputs but were not using it. Can be found and enabled like that.

File: pkgs/by-name/mp/mpb/package.nix
Comment (sheepforce) [inline pkgs/by-name/mp/mpb/package.nix:47]:
```suggestion
```
Remove that from here ...

(there are static compilation targets in nixpkgs, e.g. `pkgsStatic`, which should not be forced to build shared libraries)

File: pkgs/by-name/mp/mpb/package.nix
Comment (sheepforce) [inline pkgs/by-name/mp/mpb/package.nix:50]: And add here 
```suggestion
  ++ lib.optional (! stdenv.hostPlatform.isStatic) "--enable-shared";
```
with a conditional

**Review by sheepforce — CHANGES_REQUESTED** [review]
Thank you very much for that work, @sheevy. :blush: Happy to see the mpb support in meep. 

I've left a few suggestions for MPB and Meep directly at the corresponding lines of code. Mainly, they are some suggestions to enable the test suite and build with libctl support.

Regarding commits, we need to follow the nixpkgs commit rules. They are outlined here https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions. It seems nitpicky, but in fact the CI pipeline of ofBorg depends on well formed commit messages. The TLDR is;

  * One commit for MPB, which must read `mpb: init at 1.11.1`
  * One commit for adding MPB to meep, which could read `python3Packages.meep: add mpb support`
  * Should you want to add yourself as maintainer another commit which must be `maintainers: add sheevy`

Thank you very much for your time and effort! :sheep:

**Comment by sheevy** [comment]
Thanks for your time and patience @sheepforce. I know it would take you significantly less time to do it yourself, rather then guiding me through the process. I really appreciate it, and hope this investment will pay off.

I think I implemented all the fixes you mentioned, and nothing got lost during the squashing and rebasing. Would you mind giving it a second read and approve if you deem it up to the standard to be merged?

**Review by sheepforce — APPROVED** [review]
Thank you very much! Looks perfect to me, now! :sheep: :smiley: 

I'm always happy to help in the scientific ecosystem and having a few more contributors is certainly a good thing. :)

File: pkgs/by-name/mp/mpb/package.nix
Comment (markuskowa) [inline pkgs/by-name/mp/mpb/package.nix:66]:
```suggestion
  meta = {
    description = "MIT Photonic-Bands: computation of photonic band structures in periodic media";
    homepage = "https://mpb.readthedocs.io/en/latest/";
    license = lib.licenses.gpl2Only;
    platforms = lib.platforms.linux;
    maintainers = with lib.maintainers; [
      sheepforce
    ];
  };
```

(avoid nested `with`)

**Review by markuskowa — COMMENTED** [review]
Please fix the nested `with` statement. 
Note, I changed the title of the PR to conform with the contributing guidelines.

Otherwise LGTM.

File: pkgs/by-name/mp/mpb/package.nix
Comment (sheevy) [reply pkgs/by-name/mp/mpb/package.nix:66]: Will do. Shall I also change it in `meep/default.nix` (I based mpb derivation on that file)?

Comment (markuskowa) [reply pkgs/by-name/mp/mpb/package.nix:66]: Yes, that would be nice.  These nested `with`s are legacy construction that we are trying to clean up.

**Comment by sheevy** [comment]
Addressed the lates feedback, thanks for it!
Is there a remote chance this change will make it into NixOS 24.11?

**Comment by markuskowa** [comment]
> Is there a remote chance this change will make it into NixOS 24.11?
Yes, I will merge at soon as the CI is finished.

@ofborg build python3Packages.meep mbp

**Pull request merged**
